File: apps/runtime/src/git/repository/parse.rs
//! Parsers for the machine-readable Git output the service reads.

use super::*;

pub(super) fn fields_with_lf(bytes: &[u8], width: usize) -> AppResult<Vec<Vec<String>>> {
    let mut records = Vec::new();
    let mut rest = bytes;
    while !rest.is_empty() {
        let mut fields = Vec::with_capacity(width);
        for _ in 0..width {
            let end = rest
                .iter()
                .position(|byte| *byte == 0)
                .ok_or_else(malformed)?;
            fields.push(text(&rest[..end])?.to_owned());
            rest = &rest[end + 1..];
        }
        rest = rest.strip_prefix(b"\n").ok_or_else(malformed)?;
        records.push(fields);
    }
    Ok(records)
}

/// `git log -z --format=…%x00…` output: every field and every record end is a
/// NUL, so the whole answer is one flat NUL-separated list whose length is a
/// multiple of the record width.
///
/// It is separate from [`fields_with_lf`] because that one reads `for-each-ref`
/// output, where a record ends with a newline. Reading one with the other's
/// rule turns a subject containing a newline — which a reflog message may — into
/// a malformed record.
pub(super) fn fields_with_nul(bytes: &[u8], width: usize) -> AppResult<Vec<Vec<String>>> {
    if bytes.is_empty() {
        return Ok(vec![]);
    }
    let mut fields: Vec<_> = bytes.split(|byte| *byte == 0).collect();
    if fields.last() == Some(&b"".as_slice()) {
        fields.pop();
    }
    if width == 0 || fields.len() % width != 0 {
        return Err(malformed());
    }
    fields
        .chunks_exact(width)
        .map(|row| {
            row.iter()
                .map(|value| text(value).map(str::to_owned))
                .collect::<AppResult<Vec<String>>>()
        })
        .collect()
}

pub(super) fn parse_tracking(track: &str) -> AppResult<(Option<u64>, Option<u64>, bool)> {
    if track == "gone" {
        return Ok((None, None, true));
    }
    let mut ahead = 0;
    let mut behind = 0;
    for part in track.split(", ").filter(|part| !part.is_empty()) {
        let (direction, count) = part.split_once(' ').ok_or_else(malformed)?;
        let count = count.parse().map_err(|_| malformed())?;
        match direction {
            "ahead" => ahead = count,
            "behind" => behind = count,
            _ => return Err(malformed()),
        }
    }
    Ok((Some(ahead), Some(behind), false))
}

pub(super) fn parse_history(
    bytes: &[u8],
    refs: &HashMap<String, Vec<String>>,
) -> AppResult<Vec<CommitRecord>> {
    if bytes.is_empty() {
        return Ok(vec![]);
    }
    let mut fields: Vec<_> = bytes.split(|byte| *byte == 0).collect();
    if fields.last() == Some(&b"".as_slice()) {
        fields.pop();
    }
    if fields.len() % 7 != 0 {
        return Err(malformed());
    }
    fields
        .as_chunks::<7>()
        .0
        .iter()
        .map(|row| {
            let oid = text(row[0])?.to_owned();
            let parents: Vec<String> = text(row[1])?
                .split_whitespace()
                .map(str::to_owned)
                .collect();
            if !valid_oid(&oid) || parents.iter().any(|oid| !valid_oid(oid)) {
                return Err(malformed());
            }
            Ok(CommitRecord {
                refs: refs.get(&oid).cloned().unwrap_or_default(),
                oid,
                parents,
                subject: text(row[6])?.into(),
                author_name: text(row[2])?.into(),
                author_email: text(row[3])?.into(),
                author_time: text(row[4])?.into(),
                committer_time: text(row[5])?.into(),
            })
        })
        .collect()
}

pub(super) fn parse_worktrees(bytes: &[u8]) -> AppResult<Vec<WorktreeRecord>> {
    let mut records = Vec::new();
    let mut current: Option<WorktreeRecord> = None;
    for field in bytes.split(|byte| *byte == 0) {
        let line = text(field)?;
        if line.is_empty() {
            if let Some(record) = current.take() {
                records.push(record);
            }
            continue;
        }
        if let Some(path) = line.strip_prefix("worktree ") {
            if current.is_some() {
                return Err(malformed());
            }
            current = Some(WorktreeRecord {
                path: path.into(),
                head_oid: None,
                branch: None,
                detached: false,
                bare: false,
                is_main: records.is_empty(),
                locked: false,
                lock_reason: None,
                prunable: false,
                prune_reason: None,
                accessible: false,
                dirty: None,
            });
        } else {
            let record = current.as_mut().ok_or_else(malformed)?;
            if let Some(oid) = line.strip_prefix("HEAD ") {
                if !valid_oid(oid) {
                    return Err(malformed());
                }
                record.head_oid = (!oid.bytes().all(|byte| byte == b'0')).then(|| oid.into());
            } else if let Some(branch) = line.strip_prefix("branch ") {
                record.branch = Some(branch.strip_prefix("refs/heads/").unwrap_or(branch).into());
            } else if line == "detached" {
                record.detached = true;
            } else if line == "bare" {
                record.bare = true;
            } else if line == "locked" || line.starts_with("locked ") {
                record.locked = true;
                record.lock_reason = line.strip_prefix("locked ").map(str::to_owned);
            } else if line == "prunable" || line.starts_with("prunable ") {
                record.prunable = true;
                record.prune_reason = line.strip_prefix("prunable ").map(str::to_owned);
            } else {
                return Err(malformed());
            }
        }
    }
    if let Some(record) = current {
        records.push(record);
    }
    Ok(records)
}

File: apps/desktop/src/core/git/repository/parse.ts
import { malformed, validOid } from "../support";
import type { CommitRecord, WorktreeRecord } from "./types";

/**
 * Parsers for the machine-readable Git output the repository service reads.
 *
 * A port of the pre-merge implementation. The two record
 * splitters are separate on purpose: `for-each-ref` ends a record with a
 * newline, `git log -z` with a NUL, and reading one with the other's rule turns
 * a subject containing a newline — which a reflog message may — into a
 * malformed record.
 */

/** `for-each-ref` output: `width` NUL-terminated fields, then a newline. */
export function fieldsWithLf(bytes: Buffer, width: number): string[][] {
  const records: string[][] = [];
  let rest = bytes;
  while (rest.length > 0) {
    const fields: string[] = [];
    for (let index = 0; index < width; index += 1) {
      const end = rest.indexOf(0);
      if (end < 0) throw malformed();
      fields.push(rest.subarray(0, end).toString("utf8"));
      rest = rest.subarray(end + 1);
    }
    if (rest[0] !== 0x0a) throw malformed();
    rest = rest.subarray(1);
    records.push(fields);
  }
  return records;
}

/** `git log -z --format=…%x00…`: one flat NUL-separated list. */
export function fieldsWithNul(bytes: Buffer, width: number): string[][] {
  if (bytes.length === 0) return [];
  const fields = splitNul(bytes);
  if (width === 0 || fields.length % width !== 0) throw malformed();
  const records: string[][] = [];
  for (let index = 0; index < fields.length; index += width) {
    records.push(
      fields.slice(index, index + width).map((value) => value.toString("utf8")),
    );
  }
  return records;
}

/** NUL-split with the empty trailing field Git leaves behind dropped. */
export function splitNul(bytes: Buffer): Buffer[] {
  const fields: Buffer[] = [];
  let start = 0;
  for (let index = 0; index < bytes.length; index += 1) {
    if (bytes[index] === 0) {
      fields.push(bytes.subarray(start, index));
      start = index + 1;
    }
  }
  if (start < bytes.length) fields.push(bytes.subarray(start));
  return fields;
}

/** `%(upstream:track,nobracket)` → `(ahead, behind, gone)`. */
export function parseTracking(
  track: string,
): [number | null, number | null, boolean] {
  if (track === "gone") return [null, null, true];
  let ahead = 0;
  let behind = 0;
  for (const part of track.split(", ").filter((value) => value !== "")) {
    const space = part.indexOf(" ");
    if (space < 0) throw malformed();
    const direction = part.slice(0, space);
    const count = Number.parseInt(part.slice(space + 1), 10);
    if (Number.isNaN(count)) throw malformed();
    if (direction === "ahead") ahead = count;
    else if (direction === "behind") behind = count;
    else throw malformed();
  }
  return [ahead, behind, false];
}

/** The seven-field `git log` record the history page and previews read. */
export function parseHistory(
  bytes: Buffer,
  refs: Map<string, string[]>,
): CommitRecord[] {
  if (bytes.length === 0) return [];
  const fields = splitNul(bytes);
  if (fields.length % 7 !== 0) throw malformed();
  const records: CommitRecord[] = [];
  for (let index = 0; index < fields.length; index += 7) {
    const row = fields
      .slice(index, index + 7)
      .map((value) => value.toString("utf8"));
    const oid = row[0] as string;
    const parents = (row[1] as string).split(/\s+/).filter((v) => v !== "");
    if (!validOid(oid) || parents.some((parent) => !validOid(parent))) {
      throw malformed();
    }
    records.push({
      oid,
      parents,
      subject: row[6] as string,
      authorName: row[2] as string,
      authorEmail: row[3] as string,
      authorTime: row[4] as string,
      committerTime: row[5] as string,
      refs: refs.get(oid) ?? [],
    });
  }
  return records;
}

/** `git worktree list --porcelain -z`. */
export function parseWorktrees(bytes: Buffer): WorktreeRecord[] {
  const records: WorktreeRecord[] = [];
  let current: WorktreeRecord | undefined;
  for (const field of splitNulKeepEmpty(bytes)) {
    const line = field.toString("utf8");
    if (line === "") {
      if (current !== undefined) {
        records.push(current);
        current = undefined;
      }
      continue;
    }
    if (line.startsWith("worktree ")) {
      if (current !== undefined) throw malformed();
      current = {
        path: line.slice("worktree ".length),
        headOid: null,
        branch: null,
        detached: false,
        bare: false,
        isMain: records.length === 0,
        locked: false,
        lockReason: null,
        prunable: false,
        pruneReason: null,
        accessible: false,
        dirty: null,
      };
      continue;
    }
    if (current === undefined) throw malformed();
    if (line.startsWith("HEAD ")) {
      const oid = line.slice("HEAD ".length);
      if (!validOid(oid)) throw malformed();
      current.headOid = /^0+$/.test(oid) ? null : oid;
    } else if (line.startsWith("branch ")) {
      const branch = line.slice("branch ".length);
      current.branch = branch.startsWith("refs/heads/")
        ? branch.slice("refs/heads/".length)
        : branch;
    } else if (line === "detached") {
      current.detached = true;
    } else if (line === "bare") {
      current.bare = true;
    } else if (line === "locked" || line.startsWith("locked ")) {
      current.locked = true;
      current.lockReason = line.startsWith("locked ")
        ? line.slice("locked ".length)
        : null;
    } else if (line === "prunable" || line.startsWith("prunable ")) {
      current.prunable = true;
      current.pruneReason = line.startsWith("prunable ")
        ? line.slice("prunable ".length)
        : null;
    } else {
      throw malformed();
    }
  }
  if (current !== undefined) records.push(current);
  return records;
}

function splitNulKeepEmpty(bytes: Buffer): Buffer[] {
  const fields: Buffer[] = [];
  let start = 0;
  for (let index = 0; index < bytes.length; index += 1) {
    if (bytes[index] === 0) {
      fields.push(bytes.subarray(start, index));
      start = index + 1;
    }
  }
  fields.push(bytes.subarray(start));
  return fields;
}

/**
 * `--numstat -z` → `additions \t deletions \t path \0`. Binary files report `-`
 * for both counts, which becomes `null` rather than a misleading zero.
 */
export function parseNumstat(
  output: string,
): Map<string, [number | null, number | null]> {
  const counts = new Map<string, [number | null, number | null]>();
  const fields = output.split("\0");
  for (let index = 0; index < fields.length; index += 1) {
    const entry = fields[index] as string;
    // `diff-tree -z` leads with the commit's own object ID in a field of its
    // own; `diff` does not. Dropping a bare OID handles both.
    if (entry.trim() === "" || validOid(entry)) continue;
    const firstTab = entry.indexOf("\t");
    const secondTab = entry.indexOf("\t", firstTab + 1);
    if (firstTab < 0 || secondTab < 0) continue;
    const additions = Number.parseInt(entry.slice(0, firstTab), 10);
    const deletions = Number.parseInt(entry.slice(firstTab + 1, secondTab), 10);
    let path = entry.slice(secondTab + 1);
    if (path === "") {
      // Rename: the two paths follow as separate NUL fields.
      index += 2;
      const destination = fields[index];
      if (destination === undefined) continue;
      path = destination;
    }
    counts.set(path, [
      Number.isNaN(additions) ? null : additions,
      Number.isNaN(deletions) ? null : deletions,
    ]);
  }
  return counts;
}

/**
 * `--name-status -z` → `<code> \0 <path> \0`, with renames emitting
 * `R100 \0 <old> \0 <new> \0`.
 */
export function parseNameStatus(
  output: string,
  normalize: (code: string) => string,
): [string, string][] {
  const entries: [string, string][] = [];
  const fields = output.split("\0").filter((field) => field !== "");
  for (let index = 0; index < fields.length; index += 1) {
    const code = fields[index] as string;
    // A status code is never 40 hex characters, so skipping a bare OID is
    // unambiguous — and reading it as a code would desynchronize every pair.
    if (validOid(code)) continue;
    const renamed = code.startsWith("R") || code.startsWith("C");
    index += 1;
    let path = fields[index];
    if (path === undefined) throw malformed();
    if (renamed) {
      index += 1;
      path = fields[index];
      if (path === undefined) throw malformed();
    }
    entries.push([normalize(code), path.replace(/\\/g, "/")]);
  }
  return entries;
}
